### DISCUSSION TITLE
[Feature] More Handlebars variables and funcions

### DISCUSSION THREAD

**Discussion by Thundercraft5** [coddingtonbear/obsidian-web#150] [Ideas]
Could there be handlebars variables and functions be added?
- `{{page.domain}}`: Variable for URL domain `domain.site.com`
- `{{page.path}}`: Variable for URL path: `/path/to/page` 
- `{{page.query}}`: Variable for URL query: `?search=query`
- `{{replace X Y}}`: Replace `X` with `Y`
- `{{regreplace X Y}}`: Replace `X` with `Y` (Regex)

**Comment by coddingtonbear** [comment]
Let's discuss these in two different groups:

# Your page-level details

There is a [built-in Web API for getting properties of a URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) -- so we could possibly have a handlebars function that retrieved various properties from your existing `page.url` variable _or_ just attach a `URL()` instance to the `page` object under some kind of reasonable name...unfortunately the name `url` is already taken, though, so !

If you were to create a PR doing either of those, I'd be up for reviewing it.

# Replacement Functions

I'm not opposed to either of these in general, and would probably approve a PR adding either of these (but please do post PRs separately from one another whenever at all possible), but I am very curious about the use cases you're trying to solve for with such functionality and would want ot hear a little more before merging anything.

**Comment by Thundercraft5** [comment]
I'm still quite new to obsidian, so I don't know exactly what all of my use cases could be for Replacement Functions. Currently, a couple use cases for replacement functions would for custom-formatting a note title (as some characters used in titles are not allowed in obsidian), for example I would want to format a title to be an acceptable filename by replacing `/` with ` - `.

**Reply by coddingtonbear** [reply]
In that case, you may just want to use [the existing `filename` helper function](https://github.com/coddingtonbear/obsidian-web/wiki/Understanding-Templates#user-content-helpers) that already strips out such characters.

**Reply by Thundercraft5** [reply]
The issue is sometimes `filename` leaves behind a badly formatted title, something like this happens with a note title: `GitHub - user/repo | A repo for something` => `GitHub - userrepo  A repo for something`

**Reply by coddingtonbear** [reply]
That sounds like maybe a situation in which you could propose and then contribute changes to that function to make the formatted title "better" instead of creating a new bespoke function, no?

**Reply by Thundercraft5** [reply]
It could be done that way, but I believe it would better to allow the user to handle that with Replacement Functions instead of trying to cover each and every case that arose.

**Reply by coddingtonbear** [reply]
I'm afraid I disagree!  If the goal is to make a particular bit of text from the document into a filename, I don't think its all that reasonable to expect that the user both know the complete (and often long!) lists of characters that are invalid on their operating system for use in filenames _and_ also know how to correctly build a regular expression for stripping out those invalid characters.  It seems like just making the existing `filename` handlebars helper do that for them is almost certainly the right choice from the user's perspective.

If you can think of another reason that a regex or replacement helper could be used, though, I'm up for hearing about it.

**Comment by kylek29** [comment]
One thing that could be done to facilitate this and add more functionality is to incorporate the Handlebars-Helpers add-on ( https://github.com/helpers/handlebars-helpers#install ). It adds both URL and match/replace functions. 

More modern fork: https://github.com/jaredwray/fumanchu
